terminology/terminology_processor.bal

```diff
@@ -0,0 +1,1661 @@
+        r4:FHIRError[] errors = [];
+        foreach i4:ValueSet valueSet in valueSets {
+            r4:FHIRError? result = self.addValueSet(valueSet);
+            _ = result is r4:FHIRError ? errors.push(result) : "";
+        }
+        return errors;
+    }
+
+    # Add list of new ValueSet as a json or json array.
+    #
+    # + valueSetJsonArray - Json ValueSet data in the JSON format
+    # + return - Return List of FHIRErrors if any
+    public isolated function addValueSetsAsJson(json[] valueSetJsonArray) returns r4:FHIRError[]? {
+        r4:FHIRError[] errors = [];
+        i4:ValueSet[] valueSets = [];
+
+        foreach var v in valueSetJsonArray {
+            i4:ValueSet|error result = v.cloneWithType(i4:ValueSet);
+            if result is error {
+                r4:FHIRError er = r4:createFHIRError(
+                    string `Invalid data. Can not parse the provided json data: ${v.toBalString()}`,
+                    r4:ERROR,
+                    r4:INVALID,
+                    diagnostic = "Please check the provided json structure and re-try with this data again",
+                    cause = result,
+                    httpStatusCode = http:STATUS_BAD_REQUEST);
+
+                errors.push(er);
+            } else {
+                valueSets.push(result);
+            }
+        }
+
+        _ = valueSets.length() > 0 ? errors.push(...<r4:FHIRError[]>self.addValueSets(valueSets)) : "";
+
+        if errors.length() > 0 {
+            return errors;
+        }
+        return;
+    }
+
+    # Add a new CodeSystem.
+    #
+    # + codeSystem - ValueSet to be added, data in the Ballerina record format
+    # + return - Return FHIRError
+    public isolated function addCodeSystem(i4:CodeSystem codeSystem) returns r4:FHIRError? {
+        lock {
+            if codeSystem.url == () {
+                return r4:createFHIRError(
+            string `Can not find the URL of the CodeSystem with name: ${codeSystem.name.toBalString()}`,
+            r4:ERROR,
+            r4:INVALID_REQUIRED,
+            diagnostic = string `Add a proper URL for the resource: 
+                                    http://hl7.org/fhir/R4/codesystem-definitions.html#CodeSystem.url`,
+            errorType = r4:VALIDATION_ERROR,
+            httpStatusCode = http:STATUS_BAD_REQUEST
+            );
+            }
+
+            if codeSystem.'version == () {
+                return r4:createFHIRError(
+            string `Can not find the version of the CodeSystem with name: ${codeSystem.name.toBalString()}`,
+            r4:ERROR,
+            r4:INVALID_REQUIRED,
+            diagnostic = string `Add appropriate version for the resource: 
+                                    https://hl7.org/fhir/R4/codesystem-definitions.html#CodeSystem.version`,
+            errorType = r4:VALIDATION_ERROR,
+            httpStatusCode = http:STATUS_BAD_REQUEST
+            );
+            }
+
+            string url = <string>codeSystem.url;
```

Comment:
Can this cast fail?

Reply:
No, we are checking null values beforehand